browser/ui/webui/chat_ui/chat_ui_page_handler.cc

```diff
@@ -0,0 +1,187 @@
+        return turn.visibility == ConversationTurnVisibility::HIDDEN;
+      });
+
+  std::transform(history.begin(), new_end_it, std::back_inserter(list),
+                 [](const ConversationTurn& turn) { return turn.Clone(); });
+
+  std::move(callback).Run(std::move(list));
+}
+
+void ChatUIPageHandler::GetSummary() {
+  active_chat_tab_helper_->GetArticleSummaryString(base::BindOnce(
+      &ChatUIPageHandler::OnArticleSummaryResult, base::Unretained(this)));
+}
+
+void ChatUIPageHandler::MarkAgreementAccepted() {
+  profile_->GetOriginalProfile()->GetPrefs()->SetBoolean(
+      chat_ui::prefs::kBraveChatHasSeenDisclaimer, true);
+}
+
+void ChatUIPageHandler::OnArticleSummaryResult(const std::u16string& result,
+                                               bool is_from_cache) {
+  if (result.empty()) {
+    page_.get()->OnContentSummarizationFailed();
+    return;
+  }
+
+  if (is_from_cache) {
+    active_chat_tab_helper_->AddToConversationHistory({
+        CharacterType::ASSISTANT,
+        ConversationTurnVisibility::VISIBLE,
+        base::UTF16ToUTF8(result),
+    });
+    page_.get()->OnConversationHistoryUpdate();
+    return;
+  }
+
+  std::string summarize_prompt = base::ReplaceStringPlaceholders(
+      l10n_util::GetStringUTF8(IDS_SUMMARIZE_PROMPT_TEMPLATE),
+      {base::UTF16ToUTF8(result)}, nullptr);
+
+  // We hide the prompt with article content from the user
+  MakeAPIRequestWithConversationHistoryUpdate(
+      {CharacterType::HUMAN, ConversationTurnVisibility::HIDDEN,
+       summarize_prompt});
+}
+
+void ChatUIPageHandler::MakeAPIRequestWithConversationHistoryUpdate(
+    ConversationTurn turn) {
+  active_chat_tab_helper_->AddToConversationHistory(turn);
+  page_.get()->OnConversationHistoryUpdate();
+
+  std::string prompt_with_history =
+      base::StrCat({active_chat_tab_helper_->GetConversationHistoryAsString(),
+                    chat_ui::kAIPrompt});
+
+  DCHECK(api_helper_);
+
+  // Assuming a hidden conversation has a summary prompt,
+  // the incoming response is expected to include the AI-generated summary.
+  // TODO(nullhook): Improve this heuristic, as it may or may not be true.
+  bool contains_summary =
+      turn.visibility == ConversationTurnVisibility::HIDDEN ? true : false;
+
+  page_.get()->OnGeneratingContent(true);
+  api_helper_->QueryPrompt(
+      base::BindOnce(&ChatUIPageHandler::OnAPIResponse, base::Unretained(this),
+                     contains_summary),
+      std::move(prompt_with_history));
+}
+
+void ChatUIPageHandler::OnAPIResponse(bool contains_summary,
+                                      const std::string& assistant_input,
+                                      bool success) {
+  if (!success) {
+    return;
+  }
+
+  ConversationTurn turn = {CharacterType::ASSISTANT,
+                           ConversationTurnVisibility::VISIBLE,
+                           assistant_input};
+  active_chat_tab_helper_->AddToConversationHistory(turn);
```

Comment:
There are some asynchronous problems here. For example, if the API request is sent and the active tab changes before the response is received, then the conversation or the article summary will be added to the incorrect tab helper. For this reason and the fact that the TabHelper
There are some asynchronous problems here. For example, if the API request is sent and the active tab changes before the response is received, then the conversation or the article summary will be added to the incorrect tab helper. For this reason and the fact that the TabHelper
There are some asynchronous problems here. For example, if the API request is sent and the active tab changes before the response is received, then the conversation or the article summary will be added to the incorrect tab helper.
The Conversation is tied to a WebContents, via the TabHelper. Therefore the TabHelper, or something owned by it, should make the requests, or handle the responses, and/or cancel when the WebContents is destroyed.

Alternatively, the TabHelper instance should be remembered when the API response comes in, but maintaining that reference could be more complicated and I'm not sure it "feels" right to have the PageHandler doing the request logic.

Reply:
This is an important edge case. It makes sense for the tab helper to own the API. This can isolate the requests to specific tab instances rather than a singleton page handler.

I have updated the code to accommodate this.